File: Data Structures & Algorithms/add-two-numbers/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1 == null || (l1.val == 0 && l1.next == null)) {
            return l2;
        }
        if (l2 == null || (l2.val == 0 && l2.next == null)) {
            return l1;
        }
        ListNode carry = new ListNode(0);
        int sum = l1.val + l2.val;
        if (sum >= 10) {
            sum = sum - 10;
            carry = new ListNode(1);
        }
        return new ListNode(sum, addTwoNumbers(l1.next, addTwoNumbers(carry, l2.next)));
    }
}

File: Data Structures & Algorithms/plus-one/submission-2.cpp
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int carry = 1;
        int size = digits.size();
        for (int i = size - 1; i >= 0; --i) {
            digits[i] += carry;
            cout << digits[i] << '\n';
            if (digits[i] == 10) {
                digits[i] = 0;
                carry = 1;
            } else {
                carry = 0;
            }
            cout << digits[i] << '\n';
        }
        if (carry == 1) {
            digits.insert(digits.begin(), carry);
        }
        return digits;
    }
};
